### PR TITLE
Features/logger

### PR DESCRIPTION
**Pull request by koparasy**
Provide environment variables to log data into files and setup verbosity.

### CODE REVIEW

File: tools/AMSLogReader.py [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+    if args.csv:
+        df.to_csv(args.out_file)
+
+    if args.plot:
+        plot(df, args.uq, args.memory, args.out_file)
+
+
+if __name__ == "__main__":
+    main()
+    main()
```
Comment (koparasy) [inline tools/AMSLogReader.py]: Reminder to delete the additional "main".

File: src/AMSlib/wf/logger.cpp [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+{
+namespace util
+{
+
+// By default AMS prints only errors
+static LogVerbosityLevel defaultLevel = LogVerbosityLevel::Error;
+
+const char* Logger::MessageLevelName[LogVerbosityLevel::Num_Levels] = {"ERROR",
+                                                                       "WARNIN"
+                                                                       "G",
```
Comment (lpottier) [inline src/AMSlib/wf/logger.cpp:48]: is that normal? The "G" alone

**Comment by koparasy** [comment]
@lpottier do you know why our tests did not catch the MPI=Off case? Should I add such a case in our tests?

**Comment by lpottier** [comment]
> @lpottier do you know why our tests did not catch the MPI=Off case? Should I add such a case in our tests?

I was wondering the same thing. I checked and there is no tests with `WITH_MPI=Off` in [ci.yml](https://github.com/LLNL/AMS/blob/develop/.github/workflows/ci.yml). We should add at least one.

**Comment by koparasy** [comment]
> > @lpottier do you know why our tests did not catch the MPI=Off case? Should I add such a case in our tests?
> 
> I was wondering the same thing. I checked and there is no tests with `WITH_MPI=Off` in [ci.yml](https://github.com/LLNL/AMS/blob/develop/.github/workflows/ci.yml). We should add at least one.

I will add one. To make sure this is covered.

File: src/AMSlib/wf/logger.cpp [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+             ams_out == nullptr,
+             "Could not open file for stdout redirection");
+    }
+  }
+}
+
+void Logger::flush()
+{
+  if (ams_out != nullptr && ams_out != stdout) fflush(ams_out);
+  fflush(ams_err);
```
Comment (lpottier) [inline src/AMSlib/wf/logger.cpp:125]: We should check that `ams_err` is not null here.

File: src/AMSlib/AMS.cpp [hunk 4]
```diff
@@ -323,9 +344,92 @@ class AMSWrap
...
+      if (ams_logger_dir) {
+        log_dir = std::string(ams_logger_dir);
+      }
+
+      char hostname[HOST_NAME_MAX];
+      if (gethostname(hostname, HOST_NAME_MAX) != 0) {
+        FATAL(AMS, "Get hostname returns error");
+      }
+
+      CFATAL(AMS, !path_exists(log_dir), "Log Directory does not exist");
```
Comment (lpottier) [inline src/AMSlib/AMS.cpp]: We already have a FATAL in `path_exists`. Also, if the `AMS_LOG_DIR` points to a non-existing directory the code segfault

**Review by lpottier — CHANGES_REQUESTED** [review]
It looks good, we just need to fix that segfault when `AMS_LOG_DIR` is pointing to a directory that does not exist.

**Review by lpottier — APPROVED** [review]

File: src/AMSlib/wf/logger.cpp (hunk 2 above)
Comment (koparasy) [reply src/AMSlib/wf/logger.cpp:48]: our clang-format does it "weird".

File: src/AMSlib/AMS.cpp (hunk 4 above)
Comment (koparasy) [reply src/AMSlib/AMS.cpp]: Fixed. We cannot use logs until we setup the loggers. I had to postpone the CFATAL condition.

File: src/AMSlib/wf/logger.cpp (hunk 3 above)
Comment (koparasy) [reply src/AMSlib/wf/logger.cpp:125]: Fixed

**Pull request merged**